src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs

```diff
@@ -116,10 +117,141 @@ private void Visit(IAssemblySymbol assembly)
             return namespaceNode;
         }
 
+        private static bool walkTypeSymbol(ITypeSymbol ty, HashSet<ITypeSymbol> visited, Func<ITypeSymbol, bool> f)
+        {
+            visited.Add(ty);
+
+            if (f(ty))
+            {
+                return true;
+            }
+
+            foreach(INamedTypeSymbol memberType in ty.GetTypeMembers())
+            {
+                if (!visited.Contains(memberType))
+                {
+                    if (walkTypeSymbol(memberType, visited, f))
+                    {
+                        return true;
+                    }
+                }
+            }
+            
+            return false;
+        }
+
+        private static bool IsOrContainsReferenceType(ITypeSymbol ty)
+        {
+            HashSet<ITypeSymbol> visited = new(SymbolEqualityComparer.Default);
+            return walkTypeSymbol(ty, visited, ty => ty.IsRefLikeType || ty.IsReferenceType); // assuming it covers ByReference<T>
+        }
+
+        private static bool IsOrContainsNonEmptyStruct(ITypeSymbol root)
+        {
+            HashSet<ITypeSymbol> visited = new(SymbolEqualityComparer.Default);
+
+            return walkTypeSymbol(root, visited, ty => {
+                if (ty.IsUnmanagedType)
```

Comment:
This entire method could use comments as to why it means "non-empty".  I know we had questions around how the old code in CCI was determining empty vs not and we had to do archeaology of code / commits / pr comments to understand that.  Let's do better here by commenting the code to explain what it's doing and why.